### PR TITLE
ARROW-13532: [C++][Compute]  - adding set membership type filtering to hash table interface

### PR DESCRIPTION
**Pull request by michalursa**
This change contains some refactoring of the code of the core hash table implementation used in grouper.
The goal of this change is to separate read-only operations on the hash table from operations modifying it.

Originally the only provided operation for hash table access was the map operation, that would return ids of matching keys found in the hash table or automatically insert new keys if they are not found assigning them new ids.

The change splits the map operation into a pipeline consisting of three stages:
- early filter
- find
- map_new_keys
The three of them called one after another provide the functionality of the map. The output of each of them is used as the input to the next one. Each next stage in the pipeline can potentially process a smaller subset of rows than the previous stage, because of filtering of the rows done at each stage. 

Early filter corresponds to the part that can be seen as an equivalent of Bloom filter. It quickly, based on hash values only and without executing any key comparison, marks the keys that definitely do not have a match in the hash table. False positives are possible, but as with the Bloom filter, their probability should be small. 

The next stage in the pipeline, find method, correspond to the full processing of all of the input rows with keys that are already present in the hash table. It is a read-only operation on the hash table. It finishes filtering from early filter getting rid of any potential false positives. It also outputs corresponding group ids for all keys found in the hash table. The caller may ignore group ids if only the filtering part is important, but there is no meaningful performance overhead in outputting them, since they are needed anyway for executing key comparisons. 

The final stage of the pipeline is completing the pre-existing map functionality, processing all new keys from the last batch. The set of new keys is identified by the result of the previous stage - filter operation. The last stage takes care of inserting new keys, assigning them new group ids, resizing the hash table when it gets too full. The number of inserted keys may be smaller than the number of keys passed to this stage, since there may be duplicates among them.

The restructuring of the code should not only be useful for the implementation of join exec node, but it should also help in the future in implementation of shared multi-threaded access. Only the last stage of the pipeline can modify the hash table, so it is the only one that requires thread synchronization. At the same time it only processes the keys that were not present in the hash table when the processing of the exec batch started, so it can be expected in many cases to be a small fraction of all the inputs.

 There is no performance difference in my run of arrow-compute-aggregate-benchmark (on Ubuntu, clang 10.0, Intel I9-10980XE) both for AVX2 build and a build without SIMD.

### CODE REVIEW

File: cpp/src/arrow/compute/exec/key_encode.cc [hunk 1]
```diff
@@ -864,6 +864,17 @@ void KeyEncoder::EncoderBinaryPair::Encode(uint32_t offset_within_row, KeyRowArr
     EncodeImp_fn[dispatch_const](num_processed, offset_within_row, rows, col_prep[0],
                                  col_prep[1]);
   }
+
+  DCHECK(temp1->metadata().is_fixed_length);
+  DCHECK(temp1->length() * temp1->metadata().fixed_length >=
```
Comment (bkietz) [inline cpp/src/arrow/compute/exec/key_encode.cc:869]: Please use the more specific macros for DCHECK and ARROW_CHECK statements
```suggestion
  DCHECK_GE(temp1->length() * temp1->metadata().fixed_length,
```
There is a lint check against this, but it occasionally fails to parse in the presence of `operator->`

File: cpp/src/arrow/compute/exec/key_encode.cc [hunk 2]
```diff
@@ -1366,8 +1377,8 @@ void KeyEncoder::KeyRowMetadata::FromColumnMetadataVector(
   // a) Boolean column, marked with fixed-length 0, is considered to have fixed-length
   // part of 1 byte. b) Columns with fixed-length part being power of 2 or multiple of row
   // alignment precede other columns. They are sorted among themselves based on size of
-  // fixed-length part. c) Fixed-length columns precede varying-length columns when both
-  // have the same size fixed-length part.
+  // fixed-length part decreasing. c) Fixed-length columns precede varying-length columns
```
Comment (bkietz) [inline cpp/src/arrow/compute/exec/key_encode.cc:1380]: "They are sorted in decreasing order of the size of their fixed-length part"

File: cpp/src/arrow/compute/exec/key_map.cc [hunk 3]
```diff
@@ -91,100 +94,197 @@ inline void SwissTable::search_block(uint64_t block, int stamp, int start_slot,
...
+               num_group_id_bits == 32);
+
+  // Optimistically use simplified lookup involving only a start block to find
+  // a single group id candidate for every input.
+#if defined(ARROW_HAVE_AVX2)
+  int num_group_id_bytes = num_group_id_bits / 8;
+  if ((hardware_flags_ & arrow::internal::CpuInfo::AVX2) && !optional_selection) {
+    extract_group_ids_avx2(num_keys, hashes, local_slots, out_group_ids, sizeof(uint64_t),
+                           8 + 8 * num_group_id_bytes, num_group_id_bytes);
+  } else {
```
Comment (bkietz) [inline cpp/src/arrow/compute/exec/key_map.cc:157]: Nit: I think it's best to keep macro guarded blocks contained
```suggestion
    return;
  }
```

**Review by bkietz — CHANGES_REQUESTED** [review]
This looks good, just a few nits:

**Review by bkietz — CHANGES_REQUESTED** [review]
The other thing I'd expect to see here is a new method on `Grouper` for looking up IDs from an ExecBatch, something like `Result<Datum> Grouper::Find(const ExecBatch&) const;`. If you'd prefer to do that in a follow up, please open a JIRA for adding that method

**Comment by michalursa** [comment]
> The other thing I'd expect to see here is a new method on `Grouper` for looking up IDs from an ExecBatch, something like `Result<Datum> Grouper::Find(const ExecBatch&) const;`. If you'd prefer to do that in a follow up, please open a JIRA for adding that method

I created ARROW-13706.

File: cpp/src/arrow/compute/exec/key_map.cc (hunk 3 above)
Comment (michalursa) [reply cpp/src/arrow/compute/exec/key_map.cc]: changed

File: cpp/src/arrow/compute/exec/key_encode.cc (hunk 2 above)
Comment (michalursa) [reply cpp/src/arrow/compute/exec/key_encode.cc]: changed

File: cpp/src/arrow/compute/exec/key_encode.cc (hunk 1 above)
Comment (michalursa) [reply cpp/src/arrow/compute/exec/key_encode.cc]: done

**Comment by bkietz** [comment]
@michalursa needs a rebase

**Comment by nirandaperera** [comment]
@michalursa is there an update on this? #10845 depends on this.

**Review by bkietz — APPROVED** [review]
LGTM. CI failure seems to be a timeout. Will merge

**Pull request closed without merging**